types/signed_header.go

```diff
@@ -22,7 +25,70 @@ func (sH *SignedHeader) Verify(untrst header.Header) error {
 		// and is a programmer bug
 		panic(fmt.Errorf("%T is not of type %T", untrst, untrstH))
 	}
+	err := untrstH.ValidateBasic()
+	if err != nil {
+		return &header.VerifyError{
+			Reason: err,
+		}
+	}
+	sHHash := sH.Header.Hash()
+	if !bytes.Equal(untrstH.LastHeaderHash[:], sHHash) {
+		return &header.VerifyError{
+			Reason: fmt.Errorf("last header hash %v does not match hash of previous header %v", untrstH.LastHeaderHash[:], sHHash),
+		}
+	}
+	sHLastCommitHash := sH.Commit.GetCommitHash(&untrstH.Header, sH.ProposerAddress)
+	if !bytes.Equal(untrstH.LastCommitHash[:], sHLastCommitHash) {
+		return &header.VerifyError{
+			Reason: fmt.Errorf("last commit hash %v does not match hash of previous header %v", untrstH.LastCommitHash[:], sHHash),
+		}
+	}
 	return sH.Header.Verify(&untrstH.Header)
 }
 
 var _ header.Header = &SignedHeader{}
+
+// ValidateBasic performs basic validation of a signed header.
+func (h *SignedHeader) ValidateBasic() error {
+	err := h.Header.ValidateBasic()
+	if err != nil {
+		return err
+	}
+
+	err = h.Commit.ValidateBasic()
+	if err != nil {
+		return err
+	}
+
+	// Handle Based Rollup case
+	if h.Validators == nil || len(h.Validators.Validators) == 0 {
+		return nil
+	}
```

Comment:
I really don't like those lines. It's exiting fast, and IMHO is very error prone. In general, I would like to see something like:

```go
switch h.rollupType() {
	case based:
		// validation for based rollup
	case optimistic:
		// validation for optimistic rollup
	// ...
}

Reply:
For example, if `len(h.Validators.Validators) == 0` signature validation is completely skipped, and it is not clear if it's a bug or a feature.

Reply:
Yes, there is no signature verification in based rollup since there is no specified block proposer. Also, note that this code is simply moved from `validation.go` to `signed header.go` because of the suggestion here: https://github.com/rollkit/rollkit/pull/1026#discussion_r1238280719.

Created an issue for your suggestion: https://github.com/rollkit/rollkit/issues/1042 👍